airflow/models/base_reschedule_poke_operator.py

```diff
@@ -0,0 +1,218 @@
+    complete it should return True when a success criteria is met.
+
+    Optionally, ReschedulePokeOperators can override:
+    :py:meth:`process_result` to perform any operations after the success
+    criteria is met in :py:meth: `poke`
+
+    :py:meth:`poke` is executed at a time interval and succeed when a
+    criteria is met and fail if and when they time out.
+
+    :param soft_fail: Set to true to mark the task as SKIPPED on failure
+    :type soft_fail: bool
+    :param poke_interval: Time in seconds that the job should wait in
+        between each tries
+    :type poke_interval: int
+    :param timeout: Time, in seconds before the task times out and fails.
+    :type timeout: int
+
+    """
+    ui_color = '#9933ff'  # type: str
+
+    @apply_defaults
+    def __init__(self,
+                 *args,
+                 **kwargs) -> None:
+        super().__init__(mode='reschedule', *args, **kwargs)
+
+    @abstractmethod
+    def submit_request(self, context: Dict) -> Optional[Union[str, List, Dict]]:
+        """
+        This method should kick off a long running operation.
+        This method should return the ID for the long running operation if
+        applicable.
+        Context is the same dictionary used as when rendering jinja templates.
+
+        Refer to get_template_context for more context.
+
+        :returns: a resource_id for the long running operation.
+        :rtype: Optional[Union[String, List, Dict]]
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def poke(self, context: Dict) -> bool:
+        """
+        Function that the sensors defined while deriving this class should
+        override.
+        """
+        raise NotImplementedError
+
+
+    def process_result(self, context: Dict):
+        """
+        This method can optionally be overriden to process the result of a long running operation.
+        Context is the same dictionary used as when rendering jinja templates.
+
+        Refer to get_template_context for more context.
+        """
+        self.log.info('Using default process_result. Got result of %s. Done.',
+                      self.get_external_resource_id(context))
+
+    def execute(self, context: Dict) -> None:
+        # On the first execute call submit_request and set the
+        # external resource id.
+
+        # pylint: disable=no-value-for-parameter
+        task_reschedules = TaskReschedule.find_for_task_instance(context['ti'])
+        if not task_reschedules:
+            resource_id = self.submit_request(context)
+            self.set_external_resource_id(context, resource_id)
+
+        self.handle_reschedule(context)
+        resource_id = self.get_external_resource_id(context)
+        self.log.info("Calling process_result for %s.", resource_id)
+        self.process_result(context)
+        self.set_external_resource_id(context, None)
+
+    @staticmethod
+    def set_external_resource_id(context, value):
```

Comment:
I would generalize this, and maybe call it `set_state()`. Please add type annotations.